### DISCUSSION TITLE
Any way to save files to external variables?

### DISCUSSION THREAD

**Discussion by cycleuser** [zauberzeug/nicegui#175] [Q&A] [answered]
```Python 
 ui.upload(on_upload=lambda e: ui.notify(f'{len(e.files[0])} bytes'))
```

Is there any other way to save the file and path used in an ui.upload in a normal function outside that can be used in other place rather than just a lambda notifying the bytes?

for example, what may be needed in a reference:
```Python
path, file= ui.upload( )
```

**Comment by falkoschindler** [comment, marked as answer]
Sure, you can always define a "normal" function for handling events.

```python
from nicegui import ui
from nicegui.events import UploadEventArguments

def handle_upload(e: UploadEventArguments) -> None:
    name = e.names[0]
    content = e.files[0]
    print(f'{name=}, {len(content)=}', flush=True)

ui.upload(on_upload=handle_upload)

ui.run()
```

Python's lambda expressions are just syntactic shortcuts for regular functions. Although they are pretty limited, it is often nice to define short actions inline with the UI elements. More complex behavior, however, is better moved to a separate place.

**Reply by cycleuser** [reply]
Cool, the file name and content can be used outside now.
Then I gonna try to find out a way to get the path from the nicegui.events.

**Reply by falkoschindler** [reply]
I think there is no way to get the complete file path from within the browser: https://stackoverflow.com/a/15201258/3419103

**Reply by cycleuser** [reply]
Yep, there need to be some other way to record the uploaded file path...

**Comment by catbears** [comment]
The code actually gives an AttributeError: 'UploadEventArguments' object has no attribute 'names'. If I change that to 'name', it gives me an error, that it has no attribute 'files'.
I went down this road for several hours, learned about spooledtemoraryfiles and stuff, but couldn't figure it out how to upload a file, save it (or feed it to opencv, or whatever) and then show the output. Is there a working example somewhere?

**Reply by falkoschindler** [reply]
Sorry, the API of `ui.upload` has changed. More precisely, the event arguments now contain a single file with `name`, `type` and `content`: https://github.com/zauberzeug/nicegui/blob/9d34d21cf65290764476f6da5d486d67725afec3/nicegui/events.py#L66-L69
Here we show how to access the file name: https://nicegui.io/reference#file_upload

**Reply by catbears** [reply]
Thank you :)

In case someone else is looking into this, here is the code with the recent API
```python
from nicegui import ui
from nicegui.events import UploadEventArguments


def handle_upload(e: UploadEventArguments) -> None:
    name = e.name
    content = e.content
    print(f'{name=}, {len(content.read())=}', flush=True)


ui.upload(on_upload=handle_upload)

ui.run()
```

**Comment by superkeyor** [comment]
Just figure it out that `e.content` is a `SpooledTemporaryFile` object. One can save it directly to disk:
```
with open(file_path, 'wb') as file_writer:
    content = e.content.read()
    file_writer.write(content)
```
